LakeFSFS - TokenProvider AWS (no login yet)

Isan-Rivkin opened the pull request:
Closes #7537 (without call to login, waiting on #7578). 
No breaking changes. Currently not document and internal only.

Isan-Rivkin commented:
@arielshaqed it's still a draft but please take a look at the draft im sure you will have comments.
If it's not clear then i'll come back and continue.

arielshaqed commented on clients/hadoopfs/src/main/java/io/lakefs/auth/AWSLakeFSTokenProvider.java:
```diff
@@ -0,0 +1,195 @@
+    }
+
+    @Override
+    public void refresh() {
```
I don't understand, it looks like this will _never_ refresh!  The first time it calls newToken it sets lakeFSAuthToken.  And every time after that !needsNewToken, so it does nothing.

Isan-Rivkin replied:
I can see very clearly see why it's confusing sorry for that. 

1. Regardless of implementation which we can discuss - the goal of this function is to create a new token (with new TTL) if possible. 
1. I'm trying to make a standard for calling refresh token, mimicking how [AWSCredentialProvider](https://github.com/aws/aws-sdk-java/blob/1.7.4/src/main/java/com/amazonaws/auth/AWSCredentialsProvider.java#L44) did that.
1. I'm open for discussion on removing it for the initial release - e.g no refresh will be delivered.

arielshaqed replied:
So please document what it does: it clearly confuses at least one reader of the code.

arielshaqed reviewed (commented):
THANKS!

Very nice.  All code here is very good.  But I think I am missing several parts, notably I don't understand how refresh can ever **re--**fresh the token, it will only set it once.  I think we will need an additional executor / sleeping thread / something that periodically refreshes.  So not sure what parts are included in this draft, basically.

AFAICT this requires sigV4 on the AWS endpoint.  This is probably cool, but best validate versus requirements.  In particular, us-east-1 is always special with regards to sigV4/sigV2.

Isan-Rivkin commented on clients/hadoopfs/src/main/java/io/lakefs/auth/AWSLakeFSTokenProvider.java:
```diff
@@ -0,0 +1,195 @@
+import java.util.Map;
+import com.amazonaws.util.json.JSONObject;
+
+class GeneratePresignGetCallerIdentityRequest extends AmazonWebServiceRequest {
```
It does unless the class is not declared as `public` 👀

Isan-Rivkin replied:
fixed

Isan-Rivkin commented on clients/hadoopfs/src/main/java/io/lakefs/auth/AWSLakeFSTokenProvider.java:
```diff
@@ -0,0 +1,195 @@
+}
+
+class GetCallerIdentityV4Presigner extends AWS4Signer implements  STSGetCallerIdentitPresigner{
+    /**
+     * Seconds in a week, which is the max expiration time Sig-v4 accepts
+     */
+    private final static long MAX_EXPIRATION_TIME_IN_SECONDS = 60 * 60 * 24 * 7;
```
removed

Isan-Rivkin commented on clients/hadoopfs/src/main/java/io/lakefs/auth/AWSLakeFSTokenProvider.java:
```diff
@@ -0,0 +1,195 @@
+import java.util.Map;
+import com.amazonaws.util.json.JSONObject;
+
+class GeneratePresignGetCallerIdentityRequest extends AmazonWebServiceRequest {
```
THANK YOU this is unnecessary, removed it and adopted the code.

Isan-Rivkin commented on clients/hadoopfs/src/main/java/io/lakefs/Constants.java:
```diff
@@ -5,11 +5,22 @@
+    public static final String TOKEN_AWS_CREDENTIALS_PROVIDER_ACCESS_KEY_SUFFIX = "token.aws.access.key";
+    public static final String TOKEN_AWS_CREDENTIALS_PROVIDER_SECRET_KEY_SUFFIX = "token.aws.secret.key";
+    public static final String TOKEN_AWS_CREDENTIALS_PROVIDER_SESSION_TOKEN_KEY_SUFFIX = "token.aws.session.token";
+    public static final String TOKEN_AWS_CREDENTIALS_PROVIDER_TOKEN_DURATION = "token.aws.sts.duration";
```
fixed duration. 
will add documentation **before** merging this as a final step in additional PR into this one, i just want to be sure about the actual configuration

Isan-Rivkin commented:
@arielshaqed please re-review. 
I have addressed all your comments.
This PR will require additional look before final approve it depends on missing functionality PR #7578  being merged. 
Once that merged I'll use the calls in `newToken()`.

arielshaqed commented on clients/hadoopfs/pom.xml:
```diff
@@ -287,7 +287,7 @@ To export to S3:
             <groupId>io.lakefs</groupId>
             <artifactId>sdk</artifactId>
-            <version>1.12.0</version>
+            <version>1.14.0</version>
```
Why +2?

Isan-Rivkin replied:
1.14 + contains all external principal API's. 
So it will end up 1.14.x with the login API merged.

arielshaqed commented on clients/hadoopfs/src/main/java/io/lakefs/LakeFSClient.java:
```diff
@@ -21,49 +27,77 @@ public class LakeFSClient {
+            HttpBasicAuth basicAuth = (HttpBasicAuth) apiClient.getAuthentication(BASIC_AUTH);
+            basicAuth.setUsername(accessKey);
+            basicAuth.setPassword(secretKey);
+        }else {
```
nit:
```suggestion
        } else {
```

arielshaqed commented on clients/hadoopfs/src/main/java/io/lakefs/auth/AWSLakeFSTokenProvider.java:
```diff
@@ -0,0 +1,151 @@
+    int stsExpirationInSeconds;
+    ApiClient lakeFSApi;
+
+    AWSLakeFSTokenProvider() {
```
This is a very odd constructor.  It does nothing except change the scope.  Can we get rid of it?

Isan-Rivkin replied:
That's a default constructor I'm using it because the child class(es) `TemporaryAWSCredentialsLakeFSTokenProvider` initiates it and they also do some work before initiating it.

arielshaqed commented on clients/hadoopfs/src/main/java/io/lakefs/auth/GeneratePresignGetCallerIdentityRequest.java:
```diff
@@ -0,0 +1,38 @@
+package io.lakefs.auth;
```
Nit: this is a value type.  Could we use [org.immutables.value](https://immutables.github.io/)?

arielshaqed commented on clients/hadoopfs/src/main/java/io/lakefs/auth/LakeFSTokenProvider.java:
```diff
@@ -0,0 +1,8 @@
+package io.lakefs.auth;
+
+
+public interface LakeFSTokenProvider {
+    String getToken();
+
+    void refresh();
+}
```
Please document interface methods.

arielshaqed commented on clients/hadoopfs/src/test/java/io/lakefs/FSConfigurationTest.java:
```diff
@@ -46,4 +49,30 @@ public void testGetInt() {
+        conf.set("fs.lakefs.map2", "k1:v1,k2:v2,k3:v3");
+        conf.set("fs.lakefs.map3", "k1:v1,k2:v2,");
+        conf.set("fs.lakefs.map4", "");
+//        conf.set("fs.lakefs.map5", null);
```
?

arielshaqed commented on clients/hadoopfs/src/test/java/io/lakefs/auth/GetCallerIdentityV4PresignerTest.java:
```diff
@@ -0,0 +1,66 @@
+        AWSSessionCredentials awsCreds = GetCallerIdentityV4PresignerTest.newMockAWSCreds();
+        GetCallerIdentityV4Presigner stsPresigner = new GetCallerIdentityV4Presigner();
+        GeneratePresignGetCallerIdentityRequest stsReq = new GeneratePresignGetCallerIdentityRequest(
+                new URI("https://sts.amazonaws.com"),
```
This needs to be configurable.  Think AWS in China, govclouds, various regional endpoints, etc.

Isan-Rivkin replied:
Not sure what you mean? this **is** configurable, it's just a test Im initiating the address, I could put other addresses.

arielshaqed commented on clients/hadoopfs/src/test/java/io/lakefs/auth/GetCallerIdentityV4PresignerTest.java:
```diff
@@ -0,0 +1,66 @@
+        GeneratePresignGetCallerIdentityRequest stsReq = new GeneratePresignGetCallerIdentityRequest(
+                new URI("https://sts.amazonaws.com"),
+                awsCreds,
+                new HashMap<String,String>(){{
```
Can we use e.g. [ImmutableMap](https://guava.dev/releases/23.0/api/docs/com/google/common/collect/ImmutableMap.html)?

Isan-Rivkin replied:
Dependency that exists only in tests we discussed and decided not to introduce into lakeFSFS.

arielshaqed commented on clients/hadoopfs/src/test/java/io/lakefs/auth/GetCallerIdentityV4PresignerTest.java:
```diff
@@ -0,0 +1,66 @@
+
+        String pattern = "https://sts\\.amazonaws\\.com/\\?"+
+                "X-Amz-Date=\\d{8}T\\d{6}Z&"
+                +"Action=GetCallerIdentity&"
+                + "X-Amz-Algorithm=AWS4-HMAC-SHA256&"+
```
Please position "+" signs consistently.

arielshaqed commented on clients/hadoopfs/src/test/java/io/lakefs/auth/GetCallerIdentityV4PresignerTest.java:
```diff
@@ -0,0 +1,66 @@
+                "Version=2011-06-15&"+
+                "X-Amz-SignedHeaders=host%3Bx-lakefs-server-id&"+
+                "X-Amz-Security-Token=[^&]+&"+
+                "X-Amz-Credential="+ awsCreds.getAWSAccessKeyId()+"%2F\\d{8}%2Fus-east-1%2Fsts%2Faws4_request&"+
```
Ideally you would extract the credential to a group using `(\w+)`, match, and then test equality to the AWS access key ID.  The current code might behave weirdly if the access key ID is somehow invalid and contains strange characters.  We can work around that, or we can make an effort to explain why it cannot happen, or we can make the effort to quote the access key ID -- but it will still be simplest simply not to use a regexp to do that.

I know it's just a test.  OTOH, "code returned a bad access key" would be a really good error message :-)

Isan-Rivkin replied:
updated with specific check for that param and a message

arielshaqed commented on clients/hadoopfs/src/test/java/io/lakefs/auth/GetCallerIdentityV4PresignerTest.java:
```diff
@@ -0,0 +1,66 @@
+
+        Pattern compiledPattern = Pattern.compile(pattern);
+        Matcher matcher = compiledPattern.matcher(url.toString());
+        Assert.assertEquals(String.format("URL %s does not match \npattern: %s",url, pattern), true, matcher.matches());
```
This unit test verifies that the code returns all query parameters in a particular order.  But I do not think that this should be part of the requirement for what we are testing.

So I would really prefer to _parse_ the HRL, and then verify that the fields are correct.

Isan-Rivkin replied:
updated the test to check and compare each parameter one by one as suggested, thanks!

arielshaqed reviewed (approved):
Neat!

How was this tested?  Can we incorporate it into an Esti test, or into the contract test?  I worry about regressing.

Isan-Rivkin commented:
Thanks @arielshaqed addressed all your comments. 
Merging this PR and waiting for #7578 to be merged. 
Once it's merged I'll create additional PR with call to that API in the `LakeFSClient` and documentation accordingly.
